File: Java/Inheritance/ObjectClassTest/Interval.java
//a class declared without an explict superclass implicitly extends java.lang.Object class1
class Interval {

	private int min, sec;

	public Interval(int m, int s) {
		min = m + s / 60;
		sec = s % 60;
	}

	public int minutes() { return min; }

	public int seconds() { return sec; }

	public int time() { return 60 * min + sec; }

	//overriding toString method of java.lang.Object to return string representation of this object
	public String toString() {
		return String.format("%d:%02d", min, sec);
	}

	public int hashCode() {
		return min + sec;
	}

	public boolean equals(Object other) {
		if(other instanceof Interval that)
			return (this.min == that.min) && (this.sec == that.sec);
		return false;
	}

}

File: DotNet/Generics/GenTypeTest/DemoApp/Interval.cs
partial class Interval : IComparable<Interval>
{
    public int Minutes { get; private set;}

    public int Seconds { get; private set; }

    public Interval(int min, int sec)
    {
        Minutes = min + sec / 60;
        Seconds = sec % 60;
    }

    public int Time()
    {
        return 60 * Minutes + Seconds;
    }

    public override string ToString()
    {
        return string.Format("{0}:{1:00}", Minutes, Seconds);
    }

    public override int GetHashCode()
    {
        return Minutes + Seconds;
    }

    public override bool Equals(object other)
    {
        return other is Interval that
            && this.Minutes == that.Minutes
            && this.Seconds == that.Seconds;       
    }

    public int CompareTo(Interval that)
    {
        return this.Time() - that.Time();
    }
}
